rq/worker.py

```diff
@@ -621,9 +621,21 @@ def monitor_work_horse(self, job):
                 # which we don't want to catch, so we re-raise those ones.
                 if e.errno != errno.EINTR:
                     raise
+                # Send a heartbeat to keep the worker alive.
+                self.heartbeat()
 
     def _monitor_work_horse_tick(self, job):
-        _, ret_val = os.waitpid(self._horse_pid, 0)
+        while True:
```

Comment:
At the moment, we have two nested `while True` loops, one in `monitor_work_horse()` and one in `_monitor_work_horse_tick()`. We should simplify this code by moving the heartbeat and death penalty to `monitor_work_horse()`.

Reply:
`UnixSignalDeathPenalty` should only be used around `os.waitpid`. Otherwise we might end up interrupting `handle_job_failure` or or code paths further down in that function.
I will implement a cleaner loop - that keeps the alarm as close as possible to the waitpid.